specs/experimental/cannon-fault-proof-vm-mt.md

```diff
@@ -0,0 +1,511 @@
+  - [Waiting Threads](#waiting-threads)
+  - [Voluntary Preemption](#voluntary-preemption)
+  - [Forced Preemption](#forced-preemption)
+- [FPVM State](#fpvm-state)
+  - [State](#state)
+  - [State Hash](#state-hash)
+  - [Thread State](#thread-state)
+  - [Thread Hash](#thread-hash)
+  - [Thread Stack Hashing](#thread-stack-hashing)
+- [Memory](#memory)
+  - [Heap](#heap)
+- [Delay Slots](#delay-slots)
+- [Syscalls](#syscalls)
+  - [Supported Syscalls](#supported-syscalls)
+  - [Noop Syscalls](#noop-syscalls)
+- [I/O](#io)
+  - [Standard Streams](#standard-streams)
+  - [Hint Communication](#hint-communication)
+  - [Pre-image Communication](#pre-image-communication)
+    - [Pre-image I/O Alignment](#pre-image-io-alignment)
+- [Exceptions](#exceptions)
+- [Security Model](#security-model)
+  - [Compiler Correctness](#compiler-correctness)
+  - [Compiler Assumptions](#compiler-assumptions)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+This is a description of the second iteration of the Cannon Fault Proof Virtual Machine (FPVM).
+When necessary to distinguish this version from the initial implementation,
+it can be referred to as Multithreaded Cannon (MTCannon).  Like the original Cannon implementation,
+the MTCannon FPVM emulates a minimal Linux-based system running on big-endian 32-bit MIPS32 architecture.
+A lot of its behaviors are copied from Linux/MIPS with a few tweaks made for fault proofs.
+For the rest of this doc, we refer to the MTCannon FPVM as simply the FPVM.
+
+Operationally, the FPVM is a state transition function. This state transition is referred to as a _Step_,
+that executes a single instruction. We say the VM is a function $f$, given an input state $S_{pre}$, steps on a
+single instruction encoded in the state to produce a new state $S_{post}$.
+$$f(S_{pre}) \rightarrow S_{post}$$
+
+Thus, the trace of a program executed by the FPVM is an ordered set of VM states.
+
+### New Features
+
+#### Multithreading
+
+MTCannon adds support for [multithreading](https://en.wikipedia.org/wiki/Thread_(computing)).
+Thread management and scheduling are typically handled by the
+[operating system (OS) kernel](https://en.wikipedia.org/wiki/Kernel_%28operating_system%29):
+programs make thread-related requests to the OS kernel via [syscalls](https://en.wikipedia.org/wiki/System_call).
+As such, this implementation includes a few new Linux-specific thread-related [syscalls](#syscalls).
+Additionally, the [FPVM state](#fpvm-state) has been modified in order to track the set of active threads
+and thread-related global state.
+
+#### Robustness
+
+To ensure no unexpected behaviors are triggered, MTCannon will now raise an exception if unrecognized syscalls are
+encountered during program execution.
+
+## Multithreading
+
+The MTCannon FPVM rotates between threads to provide
+[concurrent execution](https://en.wikipedia.org/wiki/Concurrent_computing) rather than
```

Comment:
Maybe refer to this as [_multitasking_](https://en.wikipedia.org/wiki/Computer_multitasking) which is similar to concurrent execution, but is a more appropriate concept at this level of abstraction.